File: android/app/src/androidTest/java/com/depot/app/crypto/VectorsTest.kt
package com.depot.app.crypto

import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import org.json.JSONObject
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import org.junit.runner.RunWith

/**
 * Asserts this implementation against docs/vectors/test-vectors.json, the
 * same file web/src/crypto/vectors.test.ts checks itself against.
 *
 * These run on a device because lazysodium-android loads a native library;
 * a plain JVM unit test has no .so to bind to.
 *
 * A failure here means Android and web would derive different keys, and
 * pairing would fail in the field as an unexplained SAS mismatch. Fix the
 * implementation — only regenerate the vectors when the wire format is
 * deliberately changing, and then update both sides together.
 */
@RunWith(AndroidJUnit4::class)
class VectorsTest {

    private val vectors: JSONObject by lazy {
        val json = InstrumentationRegistry.getInstrumentation().context.assets
            .open("test-vectors.json")
            .bufferedReader()
            .use { it.readText() }
        JSONObject(json)
    }

    @Test
    fun lengthPrefixedTranscriptMatches() {
        val v = vectors.getJSONObject("transcript")
        val fields = v.getJSONArray("fields")
        val parsed = (0 until fields.length()).map { fields.getString(it).fromHex() }
        assertEquals(v.getString("expected"), buildTranscript(parsed).toHex())
    }

    @Test
    fun pairingDerivationMatches() {
        val v = vectors.getJSONObject("pairing")

        // Public keys must come back out of the fixed private scalars.
        assertEquals(v.getString("clientEkPub"), scalarMultBase(v.getString("clientEkPriv").fromHex()).toHex())
        assertEquals(v.getString("depotEkPub"), scalarMultBase(v.getString("depotEkPriv").fromHex()).toHex())

        val clientIk = identityKeyPairFromSeed(v.getString("clientIkSeed").fromHex())
        val depotIk = identityKeyPairFromSeed(v.getString("depotIkSeed").fromHex())
        assertEquals(v.getString("clientIkPub"), clientIk.publicKey.toHex())
        assertEquals(v.getString("depotIkPub"), depotIk.publicKey.toHex())

        val transcript = pairingTranscript(
            PairingTranscriptInput(
                version = v.getInt("version"),
                sessionId = v.getString("sessionId").fromHex(),
                clientEk = v.getString("clientEkPub").fromHex(),
                clientIk = clientIk.publicKey,
                depotEk = v.getString("depotEkPub").fromHex(),
                depotIk = depotIk.publicKey,
                nonce = v.getString("nonce").fromHex(),
            ),
        )
        assertEquals(v.getString("transcript"), transcript.toHex())

        val shared = ecdh(v.getString("clientEkPriv").fromHex(), v.getString("depotEkPub").fromHex())
        assertEquals(v.getString("sharedSecret"), shared.toHex())

        // Both sides of the DH must agree, or pairing silently diverges.
        val sharedOther = ecdh(v.getString("depotEkPriv").fromHex(), v.getString("clientEkPub").fromHex())
        assertEquals(shared.toHex(), sharedOther.toHex())

        val keys = deriveKeys(shared, transcript)
        assertEquals(v.getString("master"), keys.master.toHex())
        assertEquals(v.getString("kC2D"), keys.kC2D.toHex())
        assertEquals(v.getString("kD2C"), keys.kD2C.toHex())
        assertEquals(v.getString("sasSeed"), keys.sasSeed.toHex())
        assertEquals(v.getString("sas"), computeSAS(keys.sasSeed))
    }

    @Test
    fun credentialSigningMatches() {
        val v = vectors.getJSONObject("credential")
        val depotIk = identityKeyPairFromSeed(vectors.getJSONObject("pairing").getString("depotIkSeed").fromHex())

        val cred = Credential(
            depotId = v.getString("depotId"),
            clientId = v.getString("clientId"),
            issuedAt = v.getLong("issuedAt"),
            expiresAt = v.getLong("expiresAt"),
            sig = v.getString("sig"),
        )
        // Ed25519 is deterministic, so the signature itself is a known answer.
        assertTrue(verifyCredential(cred, depotIk.publicKey, now = cred.issuedAt))
        assertEquals(v.getString("depotId"), depotIk.publicKey.toBase64())

        // Expiry and tampering must both be rejected.
        assertTrue(!verifyCredential(cred, depotIk.publicKey, now = cred.expiresAt))
        assertTrue(!verifyCredential(cred.copy(clientId = "someone-else"), depotIk.publicKey, now = cred.issuedAt))
    }

    @Test
    fun reconnectSignatureMatches() {
        val p = vectors.getJSONObject("pairing")
        val v = vectors.getJSONObject("reconnect")
        val clientIk = identityKeyPairFromSeed(p.getString("clientIkSeed").fromHex())

        val transcript = reconnectTranscript(
            p.getString("clientEkPub").fromHex(),
            p.getString("depotEkPub").fromHex(),
            v.getString("challengeNonce").fromHex(),
        )
        assertEquals(v.getString("transcript"), transcript.toHex())
        assertEquals(v.getString("sig"), signReconnectResponse(clientIk.privateKey, transcript))
        assertTrue(verifyReconnectResponse(v.getString("sig"), transcript, clientIk.publicKey))
    }

    @Test
    fun derivedNoncesMatchAndStayDisjoint() {
        val v = vectors.getJSONObject("nonces")
        assertEquals(
            v.getString("chunk_d2c_t7_i3"),
            deriveChunkNonce(Direction.DEPOT_TO_CLIENT, 7, 3).toHex(),
        )
        assertEquals(
            v.getString("chunk_c2d_t0_i0"),
            deriveChunkNonce(Direction.CLIENT_TO_DEPOT, 0, 0).toHex(),
        )
        assertEquals(v.getString("ctl_c2d_counter2"), deriveCtlNonce(Direction.CLIENT_TO_DEPOT, 2).toHex())
        assertEquals(v.getString("ctl_d2c_counter0"), deriveCtlNonce(Direction.DEPOT_TO_CLIENT, 0).toHex())

        // Both frame kinds share one directional key, so the nonce spaces
        // overlapping would mean catastrophic keystream reuse.
        val seen = mutableSetOf<String>()
        for (direction in listOf(Direction.CLIENT_TO_DEPOT, Direction.DEPOT_TO_CLIENT)) {
            for (i in 0 until 64) {
                seen.add(deriveCtlNonce(direction, i.toLong()).toHex())
                seen.add(deriveChunkNonce(direction, 0, i).toHex())
                seen.add(deriveChunkNonce(direction, i, 0).toHex())
            }
        }
        assertEquals(2 * (64 + 64 + 64 - 1), seen.size)
    }

    @Test
    fun encryptedFramesMatchByteForByte() {
        val f = vectors.getJSONObject("frames")

        val c = f.getJSONObject("chunk")
        val chunkFrame = encodeChunkFrame(
            c.getString("key").fromHex(),
            Direction.DEPOT_TO_CLIENT,
            EncodedChunk(
                transferId = c.getInt("transferId"),
                chunkIndex = c.getInt("chunkIndex"),
                plaintext = c.getString("plaintext").fromHex(),
                compressed = c.getBoolean("compressed"),
            ),
        )
        assertEquals(c.getString("frame"), chunkFrame.toHex())

        val decodedChunk = decodeChunkFrame(c.getString("key").fromHex(), Direction.DEPOT_TO_CLIENT, chunkFrame)
        assertEquals(c.getString("plaintext"), decodedChunk.plaintext.toHex())

        val t = f.getJSONObject("ctl")
        val ctlFrame = encodeCtlFrame(
            t.getString("key").fromHex(),
            Direction.CLIENT_TO_DEPOT,
            t.getLong("counter"),
            t.getString("plaintext").fromHex(),
        )
        assertEquals(t.getString("frame"), ctlFrame.toHex())

        val decodedCtl = decodeCtlFrame(t.getString("key").fromHex(), Direction.CLIENT_TO_DEPOT, ctlFrame)
        assertEquals(t.getLong("counter"), decodedCtl.counter)
        assertEquals(t.getString("plaintextUtf8"), String(decodedCtl.plaintext, Charsets.UTF_8))
    }

    @Test
    fun base64RoundTripsWithoutPadding() {
        val v = vectors.getJSONObject("credential")
        // The web side strips '=' padding; these strings cross the wire
        // verbatim inside the QR payload, so the encoding must agree.
        val depotId = v.getString("depotId")
        assertTrue(!depotId.contains("="))
        assertEquals(depotId, depotId.fromBase64().toBase64())
    }
}

File: web/src/crypto/vectors.test.ts
import { readFileSync, writeFileSync } from 'node:fs'
import { describe, expect, it } from 'vitest'
import { sodium } from './sodium'
import { toBase64, toHex } from './codec'
import { buildTranscript, u8, utf8 } from './transcript'
import { computeSAS, deriveKeys, ecdh, pairingTranscript } from './derive'
import { depotChallengeBytes, reconnectTranscript, signDepotChallenge, signReconnectResponse } from './reconnect'
import { Direction, deriveChunkNonce, deriveCtlNonce, encodeChunkFrame, encodeCtlFrame } from '../transport/frame'

/** Deterministic filler so vectors are reproducible without randomness. */
function seq(n: number, start: number): Uint8Array {
  const out = new Uint8Array(n)
  for (let i = 0; i < n; i++) out[i] = (start + i * 7) & 0xff
  return out
}

const VECTORS_PATH = '../docs/vectors/test-vectors.json'

/**
 * docs/test-vectors.json is the contract between the web and Android
 * implementations: both assert against it, so a change to either one that
 * moves a derived key, a SAS or a frame byte fails here instead of at
 * integration time as an unexplained SAS mismatch.
 *
 * Regenerate deliberately, only when the wire format is meant to change:
 *   REGEN_VECTORS=1 npx vitest run src/crypto/vectors.test.ts
 */
describe('cross-implementation test vectors', () => {
  it('matches docs/test-vectors.json', async () => {
    const s = await sodium()

    // --- fixed key material -------------------------------------------
    const clientEkPriv = seq(32, 1)
    const depotEkPriv = seq(32, 100)
    const clientEkPub = s.crypto_scalarmult_base(clientEkPriv)
    const depotEkPub = s.crypto_scalarmult_base(depotEkPriv)

    const clientIkSeed = seq(32, 200)
    const depotIkSeed = seq(32, 50)
    const clientIk = s.crypto_sign_seed_keypair(clientIkSeed)
    const depotIk = s.crypto_sign_seed_keypair(depotIkSeed)

    const sessionId = seq(16, 9)
    const nonce = seq(16, 77)

    // --- pairing derivation (§3.3) ------------------------------------
    const transcript = pairingTranscript({
      version: 1,
      sessionId,
      clientEk: clientEkPub,
      clientIk: clientIk.publicKey,
      depotEk: depotEkPub,
      depotIk: depotIk.publicKey,
      nonce,
    })
    const shared = await ecdh(clientEkPriv, depotEkPub)
    const keys = await deriveKeys(shared, transcript)

    // --- credential signing bytes (§3.6) ------------------------------
    const depotId = toBase64(depotIk.publicKey)
    const clientId = toBase64(clientIk.publicKey)
    const issuedAt = 1700000000000
    const expiresAt = issuedAt + 90 * 24 * 60 * 60 * 1000
    function u64be(n: number): Uint8Array {
      const out = new Uint8Array(8)
      new DataView(out.buffer).setBigUint64(0, BigInt(n))
      return out
    }
    const credBytes = buildTranscript([utf8(depotId), utf8(clientId), u64be(issuedAt), u64be(expiresAt)])
    const credSig = s.crypto_sign_detached(credBytes, depotIk.privateKey)

    // --- reconnect (§4) -----------------------------------------------
    const challengeNonce = seq(16, 33)
    const rTranscript = reconnectTranscript(clientEkPub, depotEkPub, challengeNonce)
    const rSig = await signReconnectResponse(clientIk.privateKey, rTranscript)
    const dBytes = depotChallengeBytes(clientEkPub, depotEkPub, challengeNonce)
    const dSig = await signDepotChallenge(depotIk.privateKey, clientEkPub, depotEkPub, challengeNonce)

    // --- frames (§5.3) ------------------------------------------------
    const chunkPlain = seq(64, 5)
    const chunkFrame = await encodeChunkFrame(keys.kD2C, Direction.DepotToClient, {
      transferId: 7,
      chunkIndex: 3,
      plaintext: chunkPlain,
      compressed: false,
    })
    const ctlPlain = new Uint8Array(new TextEncoder().encode('{"type":"REQUEST_FILE"}'))
    const ctlFrame = await encodeCtlFrame(keys.kC2D, Direction.ClientToDepot, 2, ctlPlain)

    const vectors = {
      _comment:
        'Known-answer vectors shared by the web and Android implementations. ' +
        'Generated from web/src/crypto/genvectors.test.ts. If a value here changes, ' +
        'the wire format changed and BOTH implementations must be updated.',
      transcript: {
        _comment: 'protocol.md §3.3 length-prefixed concatenation',
        fields: ['00', 'aabb', '', 'ff00ff'],
        expected: toHex(buildTranscript([new Uint8Array([0]), new Uint8Array([0xaa, 0xbb]), new Uint8Array(0), new Uint8Array([0xff, 0x00, 0xff])])),
      },
      pairing: {
        _comment: 'protocol.md §3.3 key derivation and §3.4 SAS',
        clientEkPriv: toHex(clientEkPriv),
        clientEkPub: toHex(clientEkPub),
        depotEkPriv: toHex(depotEkPriv),
        depotEkPub: toHex(depotEkPub),
        clientIkSeed: toHex(clientIkSeed),
        clientIkPub: toHex(clientIk.publicKey),
        depotIkSeed: toHex(depotIkSeed),
        depotIkPub: toHex(depotIk.publicKey),
        version: 1,
        sessionId: toHex(sessionId),
        nonce: toHex(nonce),
        transcript: toHex(transcript),
        sharedSecret: toHex(shared),
        master: toHex(keys.master),
        kC2D: toHex(keys.kC2D),
        kD2C: toHex(keys.kD2C),
        sasSeed: toHex(keys.sasSeed),
        sas: computeSAS(keys.sasSeed),
      },
      credential: {
        _comment: 'protocol.md §3.6 — signing bytes are length-prefixed(depotId, clientId, u64be(issuedAt), u64be(expiresAt))',
        depotId,
        clientId,
        issuedAt,
        expiresAt,
        signingBytes: toHex(credBytes),
        sig: toBase64(credSig),
      },
      reconnect: {
        _comment: 'protocol.md §4 challenge-response',
        challengeNonce: toHex(challengeNonce),
        transcript: toHex(rTranscript),
        sig: rSig,
        depotChallengeBytes: toHex(dBytes),
        depotSig: dSig,
      },
      nonces: {
        _comment: 'protocol.md §5.3 derived nonces; the two spaces must stay disjoint',
        chunk_d2c_t7_i3: toHex(await deriveChunkNonce(Direction.DepotToClient, 7, 3)),
        chunk_c2d_t0_i0: toHex(await deriveChunkNonce(Direction.ClientToDepot, 0, 0)),
        ctl_c2d_counter2: toHex(await deriveCtlNonce(Direction.ClientToDepot, 2)),
        ctl_d2c_counter0: toHex(await deriveCtlNonce(Direction.DepotToClient, 0)),
      },
      frames: {
        _comment: 'Full encrypted frames — deterministic because nonces are derived, not random',
        chunk: {
          key: toHex(keys.kD2C),
          direction: 'DepotToClient',
          transferId: 7,
          chunkIndex: 3,
          compressed: false,
          plaintext: toHex(chunkPlain),
          frame: toHex(chunkFrame),
        },
        ctl: {
          key: toHex(keys.kC2D),
          direction: 'ClientToDepot',
          counter: 2,
          plaintext: toHex(ctlPlain),
          plaintextUtf8: '{"type":"REQUEST_FILE"}',
          frame: toHex(ctlFrame),
        },
      },
    }

    if (process.env.REGEN_VECTORS) {
      writeFileSync(VECTORS_PATH, JSON.stringify(vectors, null, 2) + '\n')
      return
    }

    const committed: unknown = JSON.parse(readFileSync(VECTORS_PATH, 'utf8'))
    expect(vectors).toEqual(committed)
    void u8
  })
})
